backend/graph.py

```diff
@@ -0,0 +1,292 @@
+journalistic tone. Combine search results together into a coherent answer. Do not \
+repeat text. Cite search results using [${{number}}] notation. Only cite the most \
+relevant results that answer the question accurately. Place these citations at the end \
+of the sentence or paragraph that reference them - do not put them all at the end. If \
+different results refer to different entities within the same name, write separate \
+answers for each entity.
+
+You should use bullet points in your answer for readability. Put citations where they apply
+rather than putting them all at the end.
+
+If there is nothing in the context relevant to the question at hand, just say "Hmm, \
+I'm not sure." Don't try to make up an answer.
+
+REMEMBER: If there is no relevant information within the context, just say "Hmm, I'm \
+not sure." Don't try to make up an answer. Anything between the preceding 'context' \
+html blocks is retrieved from a knowledge bank, not part of the conversation with the \
+user.\
+"""
+
+REPHRASE_TEMPLATE = """\
+Given the following conversation and a follow up question, rephrase the follow up \
+question to be a standalone question.
+
+Chat History:
+{chat_history}
+Follow Up Input: {question}
+Standalone Question:"""
+
+
+OPENAI_MODEL_KEY = "openai_gpt_3_5_turbo"
+ANTHROPIC_MODEL_KEY = "anthropic_claude_3_haiku"
+FIREWORKS_MIXTRAL_MODEL_KEY = "fireworks_mixtral"
+GOOGLE_MODEL_KEY = "google_gemini_pro"
+COHERE_MODEL_KEY = "cohere_command"
+
+
+class AgentState(TypedDict):
+    query: str
+    documents: list[Document]
+    messages: Annotated[list[BaseMessage], add_messages]
+
+
+def get_model(model_name: str) -> LanguageModelLike:
+    gpt_3_5 = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0, streaming=True)
+    claude_3_haiku = ChatAnthropic(
+        model="claude-3-haiku-20240307",
+        temperature=0,
+        max_tokens=4096,
+        anthropic_api_key=os.environ.get("ANTHROPIC_API_KEY", "not_provided"),
+    )
+    fireworks_mixtral = ChatFireworks(
+        model="accounts/fireworks/models/mixtral-8x7b-instruct",
+        temperature=0,
+        max_tokens=16384,
+        fireworks_api_key=os.environ.get("FIREWORKS_API_KEY", "not_provided"),
+    )
+    gemini_pro = ChatGoogleGenerativeAI(
+        model="gemini-pro",
+        temperature=0,
+        max_output_tokens=16384,
+        convert_system_message_to_human=True,
+        google_api_key=os.environ.get("GOOGLE_API_KEY", "not_provided"),
+    )
+    cohere_command = ChatCohere(
+        model="command",
+        temperature=0,
+        cohere_api_key=os.environ.get("COHERE_API_KEY", "not_provided"),
+    )
+    llm: LanguageModelLike = {
+        OPENAI_MODEL_KEY: gpt_3_5,
+        ANTHROPIC_MODEL_KEY: claude_3_haiku,
+        FIREWORKS_MIXTRAL_MODEL_KEY: fireworks_mixtral,
+        GOOGLE_MODEL_KEY: gemini_pro,
+        COHERE_MODEL_KEY: cohere_command,
+    }[model_name]
+    llm = llm.with_fallbacks(
```

Comment:
ooc why aren't we using .configurable_alternatives

Reply:
kind of a stylistic preference, i feel like it's a bit verbose to do `.configurable_alternatives` and then an extra `.with_config` just to achieve very simple lookup in this case

Reply:
don't think you'd need the with_config, model selection would happen automatically based on whatever config client passes in. you wouldn't need this get_model(model_name) function at all, you would just create the llm once

Reply:
that makes perfect sense, thank you -- updated this :)